celery/backends/redis.py

```diff
@@ -138,6 +137,16 @@ def cancel_for(self, task_id):
             self.subscribed_to.discard(key)
             self._pubsub.unsubscribe(key)
 
+    @property
+    def _pubsub(self):
```

Comment:
With the current implementation, the `_pubsub` property will never be `None`. You can view several sections in the class making this check:
- https://github.com/celery/celery/pull/5145/files#diff-accb7090ee45fffb2262cbceeabef002R135
- https://github.com/celery/celery/pull/5145/files#diff-accb7090ee45fffb2262cbceeabef002R124
- https://github.com/celery/celery/pull/5145/files#diff-accb7090ee45fffb2262cbceeabef002R92

I would suggest the following:
- Create the Redis client and PubSub instance during initialization in the [start](https://github.com/celery/celery/pull/5145/files#diff-accb7090ee45fffb2262cbceeabef002L105) method.
- Have this property fetch the attribute from the thread local: 
```python
@property
def _pubsub(self):
    return getattr(self._thread, '_pubsub', None)
```

@thedrow does that sound logical? @sihrc I believe this approach will minimize the changes and effort as well.

Reply:
That's strange. If `_pubsub` is `None` nothing happens and nothing is logged.
Is it because it wasn't started yet?

Reply:
This needs to be 'pubsub' attribute as you are setting self._thread.pubsub in the start method.

Reply:
All in all, @sihrc I have tested your changes and they cleanly fix the multithreading problem we were facing when asynchronously getting task results back from redis backend. Great stuff Chris! I'm looking forward to this being approved and hopefully Celery team does a release of 4.2.2 because a lot of people are looking for these changes!

Reply:
Nice catch, thanks @jamespeacock !

Reply:
The tests should have failed because of this error.
Please make sure they do fail.